tools/checky/src/rust_file.rs

```diff
@@ -0,0 +1,42 @@
+use std::{
+    fs,
+    io::Read,
+    path::{Path, PathBuf},
+};
+
+use ariadne::Source;
+use color_eyre::{eyre::WrapErr, Result};
+use syn::parse_file;
+
+use crate::syn_context::SynContext;
+
+#[derive(Clone, Debug)]
+pub struct RustFile {
+    pub path: PathBuf,
+    pub source_id: String,
+    pub source: Source,
+    pub file: syn::File,
+}
+
+impl RustFile {
+    pub fn try_parse(path: impl AsRef<Path>) -> Result<Self> {
+        let mut file = fs::File::open(&path)
+            .wrap_err_with(|| format!("failed to open file {}", path.as_ref().display()))?;
+        let mut buffer = String::new();
+        file.read_to_string(&mut buffer).wrap_err_with(|| {
+            format!("failed to read file {} to string", path.as_ref().display())
+        })?;
+        let path = path.as_ref().to_path_buf();
```

Comment:
I'd expect this to be cloned at the end of the function and not here, and then reborrowed

Reply:
Moved the `.to_path_buf()` at the end. Did you mean this? :thinking: